File: solutions/elixir/anagram/1/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    candidates
    |> Enum.filter(&is_anagram?(String.downcase(base), String.downcase(&1)))
  end

  defp is_anagram?(base, candidate) when base == candidate, do: false

  defp is_anagram?(base, candidate) do
    normalized_base = normalize(base)
    normalized_candidate = normalize(candidate)
    normalized_base == normalized_candidate
  end

  defp normalize(word) do
    word
    |> String.downcase()
    |> String.graphemes()
    |> Enum.sort()
  end
end

File: solutions/ocaml/anagram/1/anagram.ml
let normalize x =
  x |> String.lowercase_ascii |> String.to_seq |> List.of_seq
  |> List.sort Char.compare |> List.to_seq |> String.of_seq

let is_anagram t c =
  let is_identitcal = String.lowercase_ascii t = String.lowercase_ascii c in
  let is_anagram = normalize t = normalize c in
  is_anagram && not is_identitcal

let anagrams target candidates = candidates |> List.filter (is_anagram target)
